File: smartcontract/oracle.py
from boa.interop.Ontology.Contract import Migrate
from boa.interop.System.Storage import GetContext, Get, Put, Delete
from boa.interop.System.Runtime import CheckWitness, GetTime, Notify, Serialize, Deserialize
from boa.interop.System.ExecutionEngine import GetExecutingScriptHash, GetScriptContainer
from boa.interop.Ontology.Native import Invoke
from boa.interop.System.Transaction import GetTransactionHash
from boa.builtins import *
from ontology.interop.Ontology.Runtime import Base58ToAddress

######################### Global info ########################
Admin = Base58ToAddress('AMAx993nE6NEqZjwBssUfopxnnvTdob9ij')
ONGAddress = bytearray(b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02')
Fee = "Fee"
UndoRequestKey = "UndoRequest"

"""
https://github.com/ONT-Avocados/python-template/blob/master/libs/Utils.py
"""
def Revert():
    """
    Revert the transaction. The opcodes of this function is `09f7f6f5f4f3f2f1f000f0`,
    but it will be changed to `ffffffffffffffffffffff` since opcode THROW doesn't
    work, so, revert by calling unused opcode.
    """
    raise Exception(0xF1F1F2F2F3F3F4F4)


"""
https://github.com/ONT-Avocados/python-template/blob/master/libs/SafeCheck.py
"""
def Require(condition):
    """
	If condition is not satisfied, return false
	:param condition: required condition
	:return: True or false
	"""
    if not condition:
        Revert()
    return True

def RequireWitness(witness):
    """
	Checks the transaction sender is equal to the witness. If not
	satisfying, revert the transaction.
	:param witness: required transaction sender
	:return: True if transaction sender or revert the transaction.
	"""
    Require(CheckWitness(witness))
    return True

def Main(operation, args):
    if operation == "CreateOracleRequest":
        if len(args) != 2:
            return False
        request = args[0]
        address = args[1]
        CreateOracleRequest(request, address);
    if operation == "SetOracleOutcome":
        if len(args) != 4:
            return False
        txHash = args[0]
        data = args[1]
        status = args[2]
        errMessage = args[3]
        return SetOracleOutcome(txHash, data, status, errMessage)
    if operation == "GetOracleOutcome":
        if len(args) != 1:
            return False
        txHash = args[0]
        return GetOracleOutcome(txHash)
    if operation == "SetFee":
        if len(args) != 1:
            return False
        fee = args[0]
        return SetFee(fee)
    if operation == "MigrateContract":
        if len(args) !=7:
            return False
        code = args[0]
        needStorage = args[1]
        name = args[2]
        version = args[3]
        author = args[4]
        email = args[5]
        description = args[6]
        return MigrateContract(code, needStorage, name, version, author, email, description)

def CreateOracleRequest(request, address):
    #check witness
    RequireWitness(address)

    fee = Get(GetContext(), Fee)
    #transfer ong to oracle admin address
    res = TransferONG(address, Admin, fee)
    if res == False:
        Notify(["transferONG Error"])
        return False

    #get transaction hash
    txHash = GetTransactionHash(GetScriptContainer())

    #update undoRequestMap
    undoRequestMap = GetUndoRequestMap()
    undoRequestMap[txHash] = request
    b = Serialize(undoRequestMap)
    Put(GetContext(), UndoRequestKey, b)
    Notify(["CreateOracleRequest Done", txHash, request])
    return True

def SetOracleOutcome(txHash, data, status, errMessage):
    #check witness
    RequireWitness(Admin)

    #get undoRequest map
    undoRequestMap = GetUndoRequestMap()

    #TODO : check if key exist

    #put result into storage
    result = state(data, status, errMessage)
    r = Serialize(result)
    Put(GetContext(), txHash, r)

    #remove txHash from undoRequest map
    undoRequestMap.remove(txHash)
    b = Serialize(undoRequestMap)
    Put(GetContext(), UndoRequestKey, b)
    Notify(["SetOracleOutcome Done", txHash, status, errMessage])
    return True

def GetOracleOutcome(txHash):
    v = Get(GetContext(), txHash)
    Notify(["Get oracle outcome", v])
    return v

def SetFee(fee):
    RequireWitness(Admin)
    Require(fee >= 0)
    Put(GetContext(), Fee, fee)

def MigrateContract(code, needStorage, name, version, author, email, description):
    RequireWitness(Admin)
    res = Migrate(code, needStorage, name, version, author, email, description)
    Require(res)
    Notify(["Migrate Contract successfully", Admin, GetTime()])
    return True

def GetUndoRequestMap():
    undoRequestMap = {}
    v = Get(GetContext(), UndoRequestKey)
    if len(v) != 0:
        undoRequestMap = Deserialize(v)
    return undoRequestMap

def TransferONG(fromAcct, toAcct, ongAmount):
    """
    transfer ONT
    :param fromacct:
    :param toacct:
    :param amount:
    :return:
    """
    RequireWitness(fromAcct)
    param = state(fromAcct, toAcct, ongAmount)
    res = Invoke(0, ONGAddress, 'transfer', [param])
    if res and res == b'\x01':
        return True
    else:
        return False

File: smartcontract/oracle.cs
using Ont.SmartContract.Framework.Services.Ont;
using Ont.SmartContract.Framework;
using System;
using System.ComponentModel;
using Ont.SmartContract.Framework.Services.System;
using Helper = Ont.SmartContract.Framework.Helper;

namespace Ont.SmartContract
{
    public class Oracle : Framework.SmartContract
    {
        public static readonly byte[] admin = "AMAx993nE6NEqZjwBssUfopxnnvTdob9ij".ToScriptHash();

        public class Result
        {
            public byte[] data;
            public string status;
            public string errMessage;
        }

        public static object Main(string operation, params object[] args)
        {
            switch (operation)
            {
                case "CreateOracleRequest":
                    return CreateOracleRequest((string)args[0], (byte[])args[1]);
                case "SetOracleOutcome":
                    return SetOracleOutcome((byte[])args[0], (byte[])args[1], (string)args[2], (string)args[3]);
                case "GetOracleOutcome":
                    return GetOracleOutcome((byte[])args[0]);
                case "Migrate":
                    return Migrate((byte[])args[0], (bool)args[1], (string)args[2], (string)args[3], (string)args[4], (string)args[5], (string)args[6]);
                default:
                    return false;
            }
        }

        public static bool CreateOracleRequest(string request, byte[] address)
        {
            //TODO: check request format

            //TODO: transfer ong

            Transaction tx = (Transaction)ExecutionEngine.ScriptContainer;
            byte[] txHash = tx.Hash;
            //Runtime.Notify("txHash is :", txHash);

            Map<byte[], string> undoRequest = new Map<byte[], string>();
            byte[] v = Storage.Get(Storage.CurrentContext, "UndoRequest");
            if (v.Length != 0) {
                undoRequest = (Map<byte[], string>)Helper.Deserialize(v);
            }
            undoRequest[txHash] = request;

            byte[] b = Helper.Serialize(undoRequest);
            Storage.Put(Storage.CurrentContext, "UndoRequest", b);
            Runtime.Notify("CreateOracleRequest Done");
            return true;
        }

        public static bool SetOracleOutcome(byte[] txHash, byte[] data, string status, string errMessage)
        {
            //check witness
            if (!Runtime.CheckWitness(admin))
            {
                Runtime.Notify("Checkwitness failed.");
                return false;
            }

            Result result = new Result();
            result.data = data;
            result.status = status;
            result.errMessage= errMessage;
            byte[] r = Helper.Serialize(result);
            Storage.Put(Storage.CurrentContext, txHash, r);

            //remove txHash from undoRequest map
            Map<byte[], string> undoRequest = new Map<byte[], string>();
            byte[] v = Storage.Get(Storage.CurrentContext, "UndoRequest");
            if (v.Length != 0) {
                undoRequest = (Map<byte[], string>)Helper.Deserialize(v);
            }
            undoRequest.Remove(txHash);

            byte[] b = Helper.Serialize(undoRequest);
            Storage.Put(Storage.CurrentContext, "UndoRequest", b);
            Runtime.Notify("SetOracleOutcome Done");
            return true;
        }

        public static byte[] GetOracleOutcome(byte[] txHash)
        {
            byte[] v = Storage.Get(Storage.CurrentContext, txHash);

            //TODO: remove txHash from results
            Runtime.Notify("Get oracle outcome:", v);
            return v;
        }

        public static bool Migrate(byte[] code, bool need_storage, string name, string version, string author, string email, string description)
        {
            //check if owner of the contract
            if (!Runtime.CheckWitness(admin))
            {
                Runtime.Notify("Checkwitness failed.");
                return false;
            }

            Ont.SmartContract.Framework.Services.Ont.Contract.Migrate(code, need_storage, name, version, author, email, description);
            return true;
        }

    }
}